feat: genesis and peers flags for starport network chain show command

lubtd opened the pull request.

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,62 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+		fmt.Print(string(genesis))
+	case showPeers:
+		// Show the peers in the config.toml format
+		fmt.Printf("persistent_peers = \"%s\"", strings.Join(info.Peers, ","))
```
```suggestion
		fmt.Printf(`persistent_peers = "%s"`, strings.Join(info.Peers, ","))
```

ilgooz commented on starport/services/networkbuilder/networkbuilder.go:
```diff
@@ -456,6 +458,48 @@ func (b *Builder) StartChain(ctx context.Context, chainID string, flags []string
 
+// Generate the genesis from the launch information in a temporary directory and return this directory
+func (b *Builder) GenerateTemporaryGenesis(ctx context.Context, chainID string, homeDir string, launchInfo spn.LaunchInformation) (string, error) {
+	tmpHome, err := ioutil.TempDir("", chainID+"*")
```
I think there is no need for a chain id prefix in the temp dir.
```suggestion
	tmpHome, err := os.MkdirTemp("", "")
```

Also: https://golang.org/doc/go1.16#ioutil

ilgooz commented on starport/services/networkbuilder/networkbuilder.go:
```diff
@@ -456,6 +458,48 @@ func (b *Builder) StartChain(ctx context.Context, chainID string, flags []string
+			return "", err
+		}
+	}
+	if err := copy.Copy(originHome, tmpHome); err != nil {
```
Couldn't get why this step is needed since `chain.HomePath(tmpHome)` option already passed to `chain.New`?

lubtd replied:
`chain.HomePath(tmpHome)` is passed to `chain.New` to run the genesis generation commands inside the temporary directory but this temporary directory doesn't' have the initial home config at first

ilgooz replied:
I think is somehow relavent with https://github.com/tendermint/starport/pull/896#discussion_r599697287?

ilgooz commented on starport/services/networkbuilder/networkbuilder.go:
```diff
@@ -456,6 +458,48 @@ func (b *Builder) StartChain(ctx context.Context, chainID string, flags []string
+	}
+
+	// Run the commands to generate genesis
+	if err := generateGenesis(ctx, chainInfo, launchInfo, chainHandler); err != nil {
```
```suggestion
	err = generateGenesis(ctx, chainInfo, launchInfo, chainHandler);
	return tmpHome, err
```

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,62 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+	switch {
+	case showGenesis:
+		// Generate the genesis in a temporary directory and show the content
+		tmpHome, err := nb.GenerateTemporaryGenesis(cmd.Context(), chainID, home, info)
```
Why `home` needs to be configrable through flags?

lubtd replied:
Because the home config of the chain can be in a custom directory. Every starport command has the `--home` flag to customize the home directory

ilgooz replied:
Yeah, but from what I understand from this; in order to these flags work, one of the join or create commands needs to run first?

Because `show` is about retrieving read-only chain info so, I think it shouldn't depend on those commands. Someone who isn't interested in creating or joining to a chain should be still able to use `show` on any existent chain.

So if show command does not depend on executing create/join commands first, a customizable `home` here is also not needed.

I think, `show` also shouldn't mutate any info inside a possibly existent home directory created by the create/join commands.

lubtd replied:
I actually didn't make attention but we need to have run `create` or `join` command before and that's actually not very convenient for a potential new validator. Also,  `chain show`  is read-only in the sense that it has't any side effect on the launch workflow but in order to generate the genesis you still need to download the source, build it and initialize a config home.

I see two solution:

- The command will download the source, build and generate config home, everything in a temporary directory that will be then deleted in order to not have any side effect on the file system

- The command will download, build the binary using the same workflow as create/join/start commands (so the source will be in `.starport/spn-chain` and you can provide the `--home` flag)

The advantage of the second is that the download and building process is only performed once, for the first one even the coordinator would need to download and build each time the command is run

The disadvantage is that someone who just want to see the genesis of different chain would keep lot of files in their filesystem.

- A third solution is to do the first solution if the chain is not yet downloaded, and if the chain is already present in `.starport/spn-chain` we do the second solution (but then we must still be able to provide `--home` flag since it can be initialized in a custom home)

lubtd replied:
@fadeev maybe you would have an opinion on this as well

fadeev replied:
`show` command needs to init a chain first to apply changes to the genesis from SPN. It's perfectly fine if it creates a home directory (in the regular location) to do so. If there is something there already (rogue `~/.gaia`), then Starport should ask if it is allowed to nuke it. `--home` makes sense, if a user wants to see a genesis of a gaia-like chain, but doesn't want to remove their own `~/.gaia`.

fadeev replied:
The third solution looks good!

ilgooz replied:
I think no matter which option we go with non of them will ever need to second download the blockchain repo if it is already cached under the `spn-chain` dir. Because caching is done by the `Init` function which is the entrypoint of `networkbuilder` service.
https://github.com/tendermint/starport/blob/74bb06ff6a005177128858781c21e9005fe9f0ad/starport/services/networkbuilder/networkbuilder.go#L155-L156

I think caching the blockchain's source code in the fs and where to locate its home directory are different problems. I believe having a customizable `--home` flag for the `show` command will just be confusing for developers and doesn't add anything in terms of functionality. 

Because what we do under the hood is to create a temprorary dir as a home dir and pass it to `chain.New` in all cases. Making the user choose a home dir or using a pre-existent one won't add anything. Re-execution of `appd init` with its siblings and building a genesis doesn't take noticeable time.  

As an additional improvement, I think we can cache blockchain binaries in the similar way that we do for their source codes. But I think it's not a blocker for this PR and once it's implemented, it'll be beneficial for the create/join commands as well.

lubtd replied:
It is not that important to me to have it or not I was mainly following the Cosmos SDK command format that has the `--home` flag for all commands.

I can removing it if you think it would be confusing to the user

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,62 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+	switch {
+	case showGenesis:
+		// Generate the genesis in a temporary directory and show the content
+		tmpHome, err := nb.GenerateTemporaryGenesis(cmd.Context(), chainID, home, info)
```
```suggestion
		home, err := nb.SimulateConfigs(cmd.Context(), chainID)
```

where `home`s type is is:
```go
type chainHome string

// to avoid manually building path string in the code blow
// `genesisPath := fmt.Sprintf("%s/config/genesis.json", tmpHome)`
func (h chainHome) GenesisPath() string
```

?

lubtd replied:
Yes it makes sense to move `config/genesis.json` out of cmd. Not sure about `SimulateConfigs` there is no simulation in this function

ilgooz replied:
I felt like this kind of a simulation of a home dir -with a genesis and other sample configs-, in terms of how it would look like in reality without actually starting the chain. (not sdk simulations or anything)

lubtd replied:
"simulation" here refers to trying to start the chain and see if there is any error

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,61 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+	case showGenesis:
+		// Generate the genesis in a temporary directory and show the content
+		tmpHome, err := nb.GenerateTemporaryGenesis(cmd.Context(), chainID, home, info)
+		defer os.RemoveAll(string(tmpHome))
```
This needs to come after the err check below.

lubtd replied:
I refactor this one, we should generate the temporary directory outside the function to correct use `defer`

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,61 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+		if err != nil {
+			return err
+		}
+		genesis, err := ioutil.ReadFile(tmpHome.GenesisPath())
```
ioutil is deprecated.

lubtd replied:
I need to subscribe to Goland feed 😅

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,61 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+		fmt.Print(string(genesis))
+	case showPeers:
+		// Show the peers in the config.toml format
+		fmt.Printf(`persistent_peers = "%s"`, strings.Join(info.Peers, ","))
```
```suggestion
		fmt.Printf(`persistent_peers = "%s"\n`, strings.Join(info.Peers, ","))
```

lubtd replied:
I don't know if we should use carriage return in order to keep the output as minimalistic as possible. What do you think @fadeev ?

fadeev replied:
Peers will be copy pasted into config.toml. if we do newlines, users will have to manually insert commas, so I suggest we stick with commas.

lubtd replied:
We have commas, the question is if we should have a newline at the end of the line

fadeev replied:
Yes, sure, we should have a carriage return.

ilgooz commented on starport/services/networkbuilder/networkbuilder.go:
```diff
@@ -456,6 +458,52 @@ func (b *Builder) StartChain(ctx context.Context, chainID string, flags []string
+type ChainHome string
+
+func (h ChainHome) GenesisPath() string {
+	return fmt.Sprintf("%s/config/genesis.json", h)
```
I think this can use `chainHandler.GenesisPath()`.

ilgooz commented on starport/services/networkbuilder/simulation.go:
```diff
@@ -29,65 +27,41 @@ const ValidatorSetNilErrorMessage = "validator set is nil in genesis and still e
+	// Generate a temporary genesis from the simulated launch information
+	b.ev.Send(events.New(events.StatusOngoing, "generating genesis"))
+	tmpHome, err := b.GenerateTemporaryGenesis(ctx, chainID, homeDir, simulatedLaunchInfo)
+	defer os.RemoveAll(string(tmpHome))
```
needs to come after the error check

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -28,25 +39,61 @@ func networkChainShowHandler(cmd *cobra.Command, args []string) error {
+		fmt.Printf(`persistent_peers = "%s"`, strings.Join(info.Peers, ","))
+	default:
+		// No flag, show the chain information and launch information in yaml format
+		chain, err := nb.ShowChain(context.Background(), chainID)
```
```suggestion
		chain, err := nb.ShowChain(cmd.Context(), chainID)
```

lubtd commented:
Now there is no longer a home flag, we use `Init` to download the source in the cache and the home is temporary and deleted at the end of the command.

I refactored a bit the temporary genesis generation method, we now have this for `verify` and `approve` command as well so that anyone can verify a proposal

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -17,36 +27,99 @@ func NewNetworkChainShow() *cobra.Command {
+
+	switch {
+	case showGenesis:
+		if err := generateAndShowGenesis(cmd, nb, chainID, info); err != nil {
```
```suggestion
		return generateAndShowGenesis(cmd.Context(), nb, chainID, info)
```

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -17,36 +27,99 @@ func NewNetworkChainShow() *cobra.Command {
+	return nil
+}
+
+func generateAndShowGenesis(cmd *cobra.Command, nb *networkbuilder.Builder, chainID string, info spn.LaunchInformation) error {
```
```suggestion
func generateAndShowGenesis(ctx context.Context, nb *networkbuilder.Builder, chainID string, info spn.LaunchInformation) error {
```

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -17,36 +27,99 @@ func NewNetworkChainShow() *cobra.Command {
+		networkbuilder.SourceChainID(),
+		networkbuilder.InitializationHomePath(tmpHome),
+	)
+	if err == context.Canceled {
```
```suggestion
	if err != nil { return err }
```

seems like we have a final canceled [check](https://github.com/tendermint/starport/blob/cf58333/starport/interface/cli/starport/main.go#L40)

ilgooz commented on starport/interface/cli/starport/cmd/network_proposal_approve.go:
```diff
@@ -42,12 +47,36 @@ func networkProposalApproveHandler(cmd *cobra.Command, args []string) error {
+		return err
+	}
+
+	blockchain, err := nb.Init(
```
I see that this one is used in three ways, can we wrap the similar logic in a func?

lubtd replied:
We can actually have the temporary dir creation in `VerifyProposal` function. For show chain command I think it's good since we should create and remove the dir inside the same function

ilgooz reviewed (commented):
Seems perfect, just pointed some tweaks!

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -17,36 +27,95 @@ func NewNetworkChainShow() *cobra.Command {
-	info, err := nb.LaunchInformation(context.Background(), chainID)
+	switch {
+	case showGenesis:
+		if err := generateAndShowGenesis(cmd.Context(), nb, chainID, info); err != nil {
```
```suggestion
		return generateAndShowGenesis(cmd.Context(), nb, chainID, info)
```

ilgooz commented on starport/interface/cli/starport/cmd/network_chain_show.go:
```diff
@@ -17,36 +27,95 @@ func NewNetworkChainShow() *cobra.Command {
+	if err != nil {
+		return err
+	}
+	fmt.Print(string(genesis))
```
```suggestion
	fmt.Println(string(genesis))
```

ilgooz reviewed (approved):
👏

ilgooz commented:
One more thing, verifying second proposal exits with error.

```
tendermint/starport $ starport network proposal verify yo 0
✔ Source code fetched
✔ Blockchain initialized
✔ genesis generated
✔ genesis correctly formatted
✔ chain can be started
Proposal(s) #0 verified 🔍✅️
tendermint/starport $ starport network proposal verify yo 1
✔ Source code fetched
✔ Blockchain initialized

Error: failed to get genesis app state from config: account cosmos1rhc96z6naumk2d2d3xytmppf2as28te9sf5fta balance not in genesis state: map[]
Usage:
  liquidityd collect-gentxs [flags]

Flags:
      --gentx-dir string   override default "gentx" directory from which collect and execute genesis transactions; default [--home]/config/gentx/
  -h, --help               help for collect-gentxs

Global Flags:
      --home string         directory for config and data (default "/home/ilker/.simapp")
      --log_format string   The logging format (json|plain) (default "plain")
      --log_level string    The logging level (trace|debug|info|warn|error|fatal|panic) (default "info")
      --trace               print out full stack trace on errors

: exit status 1

```

lubtd commented:
If you check only `1` it will fails because you can't add a validator without the corresponding account. You must run `starport network proposal verify yo 0,1`

Though we should print `Proposal(s) #1 invalid 🔍❌️` as well. I will correct it in #495 because we made some refactor with `VerificationError` on this side

fadeev reviewed (approved):
Awesome!
